test/cluster/test_node_isolation.py

```diff
@@ -50,15 +50,23 @@ async def test_banned_node_cannot_communicate(manager: ManagerClient) -> None:
     logger.info(f"Unpausing {srvs[2]}")
     await manager.server_unpause(srvs[2].server_id)
 
+    log = await manager.server_open_log(srvs[2].server_id)
+    _, matches = await log.wait_for(f"received notification of being banned from the cluster from", timeout=5)
+
+    # check if the node is already banned otherwise try to connect and this should fail
+    if (len(matches) != 0):
+        return
```

Comment:
It's usually better not to branch a test like this. It makes the test harder to understand/debug and increases the risk of flakiness (one of the branches can be executed very rarely).

Also, I suspect that `len(matches) != 0` is always true with gossip messages coming every second. This makes the test useless. Can't we keep the test unchanged and make the banned node ignore the `notify_banned` RPC through an error injection?

We can also test that a node shuts down when it learns that it's banned in a new test (or at the end of this test if possible).

Reply:
Well, what this test is testing can still happen in reality to some extent. A banned node may try to handle a request before it receives `notify_banned`. It will then receive `notify_banned` as a response to sending mutations, though, so it's not a very interesting case.

I'm fine with changing `if` to `assert`. But please adjust the rest of the test then. There is no need to create the keyspace and the docstring shouldn't mention inserts. We can also consider a new name like `test_banned_node_learns_it_is_banned`.